pandas/core/groupby/generic.py

```diff
@@ -1664,3 +1655,36 @@ def _normalize_keyword_aggregation(kwargs):
         order.append((column,
                       com.get_callable_name(aggfunc) or aggfunc))
     return aggspec, columns, order
+
+
+def _recast_datetimelike_result(result):
+    """
+    If we have date/time like in the original, then coerce dates
+    as we are stacking can easily have object dtypes here.
+
+    Parameters
+    ----------
+    result : DataFrame
+
+    Notes
+    -----
+    - Assumes Groupby._selected_obj has ndim==2 and at least one
+    datetimelike column
+    - Modifies `result` inplace
```

Comment:
why is this inplace? that seems odd

Reply:
Aside from not liking functions that do stuff inplace, why wouldn't it be?  It's (partially) inplace in the status quo.

Reply:
because we almost never like to do that unless its really upfront and in your face as its not obvious. try to use maybe_convert_objects anyhow

Reply:
changed to copy and return instead of modifying inplace.  discussed maybe_convert_objects below